### PR TITLE
(PUP-4734) acceptance: ensure service providers set/remove enable links

### PR DESCRIPTION
**Pull request by er0ck**
This change removes a duplicate test and ensures the service enable
links are set and removed correctly upon more linux platforms.  upstart
is handled in its own test.  This test handles sysV and systemd.

[skip ci]

### CODE REVIEW

**Comment by MikaelSmith** [comment]
:+1: looks like an improvement.

File: acceptance/tests/resource/service/service_enable_linux.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  step "installing httpd/apache"
+  apply_manifest_on(agent, manifest_install_httpd, :catch_failures => true)
+  # chkconfig --del will remove all rc.d symlinks for this service
+  on agent, "chkconfig --del httpd", :accept_all_exit_codes => true
+  # debian platforms using sysV put rc runlevels directly in /etc/
+  on agent, "ln -s /etc/ /etc/rc.d", :accept_all_exit_codes => true
+
+  step "ensure enabling service creates the start & kill symlinks"
+  apply_manifest_on(agent, manifest_httpd_enabled, :catch_failures => true) do
+    if (platform == 'centos' || 'el' && majrelease < 7) || platform == 'debian'
```
Comment (joshcooper) [inline acceptance/tests/resource/service/service_enable_linux.rb]: I don't think the operands are being compared like you'd expect:

```
(rdb:1) platform = 'foo'
"foo"
(rdb:2) platform == 'centos' || 'el'
"el"
```

which returns truthy even though `platform` is neither 'centos' or 'el'. I think you'll want to wrap in parens to make it explicit:

```
if ((platform == 'centos' || platform == 'el') && majrelease < 7) || platform == 'debian'
```

Or rewrite the comparison logic to be more explicit

File: acceptance/tests/resource/service/service_enable_linux.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        on agent, "test -L /etc/rc.d/rc#{runlevel}.d/#{kill_symlink} && test -f #{init_script}"
+      end
+    else
+      on agent, "test -L #{symlink_el7} && test -f #{init_script_el7}"
+    end
+  end
+
+  step "ensure disabling service removes start symlinks"
+  apply_manifest_on(agent, manifest_httpd_disabled, :catch_failures => true) do
+    if (platform == 'centos' || 'el' && majrelease < 7) || platform == 'debian'
```
Comment (joshcooper) [inline acceptance/tests/resource/service/service_enable_linux.rb]: same comment

File: acceptance/tests/resource/service/service_enable_linux.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                'ubuntu' => 'apache2', 'debian' => 'apache2'}
+
+init_script_el7   = "/usr/lib/systemd/system/httpd.service"
+symlink_el7       = "/etc/systemd/system/multi-user.target.wants/httpd.service"
+
+start_runlevels = ["2", "3", "4", "5"]
+kill_runlevels = ["0", "1", "6"]
+
+agents.each do |agent|
+  platform = agent.platform.split('-')[0]
```
Comment (joshcooper) [inline acceptance/tests/resource/service/service_enable_linux.rb]: Does the `platform` always contain a dash? Is there a beaker helper to return just the platform name minus the arch?

**Comment by er0ck** [comment]
@joshcooper updated...

**Comment by joshcooper** [comment]
This fails on debian 8

```
$ be rake ci:test:aio SHA=f8da42bdb0b7df21634bee6e9b926a82a8c6c6c7 SUITE_VERSION=1.2.3.31.gf8da42b TESTS=tests/resource/service/service_enable_linux.rb CONFIG=config/nodes/debian-8-x86_64.yaml
...
* installing httpd/apache

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:16$ mktemp -t apply_manifest.pp.XXXXXX
/tmp/apply_manifest.pp.tjT9lv

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 0.01 seconds
localhost $ scp /var/folders/2q/qhdc5tpx4knd4pjr1hdskmjc0000gn/T/beaker20150901-9280-5crpc6 agent:/tmp/apply_manifest.pp.tjT9lv {:ignore => }

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:16$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.tjT9lv
Notice: Compiled catalog for q8z5rh9ajl2ddhl.delivery.puppetlabs.net in environment production in 0.71 seconds
Info: Applying configuration version '1441147516'
Notice: /Stage[main]/Main/Package[apache2]/ensure: created
Notice: Applied catalog in 6.34 seconds

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 8.14 seconds
Exited: 2

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:25$ chkconfig --del httpd
bash: chkconfig: command not found

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 0.01 seconds
Exited: 127

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:25$ ln -s /etc/ /etc/rc.d

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 0.01 seconds

  * ensure enabling service creates the start & kill symlinks

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:25$ mktemp -t apply_manifest.pp.XXXXXX
/tmp/apply_manifest.pp.SkaaaU

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 0.01 seconds
localhost $ scp /var/folders/2q/qhdc5tpx4knd4pjr1hdskmjc0000gn/T/beaker20150901-9280-165fae3 agent:/tmp/apply_manifest.pp.SkaaaU {:ignore => }

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:25$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.SkaaaU
Notice: Compiled catalog for q8z5rh9ajl2ddhl.delivery.puppetlabs.net in environment production in 0.47 seconds
Info: Applying configuration version '1441147525'
Notice: /Stage[main]/Main/Service[apache2]/enable: enable changed 'false' to 'true'
Notice: Applied catalog in 0.13 seconds

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 1.69 seconds
Exited: 2

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) 15:45:26$ test -L /etc/rc.d/rc2.d/S85httpd && test -f /etc/init.d/apache2

q8z5rh9ajl2ddhl.delivery.puppetlabs.net (agent) executed in 0.01 seconds
Exited: 1
#<Beaker::Host::CommandFailure: Host 'q8z5rh9ajl2ddhl.delivery.puppetlabs.net' exited with 1 running:
 test -L /etc/rc.d/rc2.d/S85httpd && test -f /etc/init.d/apache2
```

**Comment by joshcooper** [comment]
Also fails on `sles-12`:

```
$ be rake ci:test:aio SHA=96ffe59c225cc9e4c34cf6c75dbca7356b1fb80e SUITE_VERSION=1.2.3.30.g96ffe59 TESTS=tests/resource/service/service_enable_linux.rb CONFIG=config/nodes/sles-12-x86_64.yaml
...
  * installing httpd/apache

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) 15:52:38$ mktemp -t apply_manifest.pp.XXXXXX
/tmp/apply_manifest.pp.IxPQ2F

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) executed in 0.03 seconds
localhost $ scp /var/folders/2q/qhdc5tpx4knd4pjr1hdskmjc0000gn/T/beaker20150901-9375-12ltchd agent:/tmp/apply_manifest.pp.IxPQ2F {:ignore => }

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) 15:52:38$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.IxPQ2F
Notice: Compiled catalog for bh8ud743ne6tgub.delivery.puppetlabs.net in environment production in 0.73 seconds
Info: Applying configuration version '1441147959'
Notice: /Stage[main]/Main/Package[httpd]/ensure: created
Notice: Applied catalog in 3.69 seconds

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) executed in 5.44 seconds
Exited: 2

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) 15:52:44$ chkconfig --del httpd
httpd: unknown service

Note: This output shows SysV services only and does not include native
systemd services. SysV configuration data might be overridden by native
systemd configuration.

If you want to list systemd services use 'systemctl list-unit-files'.
To see services enabled on particular target use
'systemctl list-dependencies [target]'.


bh8ud743ne6tgub.delivery.puppetlabs.net (agent) executed in 0.17 seconds
Exited: 1

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) 15:52:44$ ln -s /etc/ /etc/rc.d

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) executed in 0.03 seconds

  * ensure enabling service creates the start & kill symlinks

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) 15:52:44$ mktemp -t apply_manifest.pp.XXXXXX
/tmp/apply_manifest.pp.e8YWSZ

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) executed in 0.03 seconds
localhost $ scp /var/folders/2q/qhdc5tpx4knd4pjr1hdskmjc0000gn/T/beaker20150901-9375-1rjizd9 agent:/tmp/apply_manifest.pp.e8YWSZ {:ignore => }

bh8ud743ne6tgub.delivery.puppetlabs.net (agent) 15:52:44$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.e8YWSZ
Notice: Compiled catalog for bh8ud743ne6tgub.delivery.puppetlabs.net in environment production in 0.48 seconds
Info: Applying configuration version '1441147965'
Error: Could not enable httpd:
Error: /Stage[main]/Main/Service[httpd]/enable: change from false to true failed: Could not enable httpd:
Notice: Applied catalog in 0.09 seconds
```

**Comment by er0ck** [comment]
we have debian 8 builds and matrix??  hmmm... looking

**Comment by MikaelSmith** [comment]
I assume Josh is running manually, Debian 8 shouldn't run except on nightly (stable).

**Comment by joshcooper** [comment]
yep in nightlies, also the pe-only agents matrix adds some more

**Comment by joshcooper** [comment]
Also ubuntu 15.04, looks like it didn't skip due to `vivid`:

```
be rake ci:test:aio SHA=f8da42bdb0b7df21634bee6e9b926a82a8c6c6c7 SUITE_VERSION=1.2.3.31.gf8da42b TESTS=tests/resource/service/service_enable_linux.rb CONFIG=config/nodes/ubuntu-1504-x86_64.yaml
...
  * installing httpd/apache

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) 15:59:51$ mktemp -t apply_manifest.pp.XXXXXX
/tmp/apply_manifest.pp.Fqrk8n

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) executed in 0.03 seconds
localhost $ scp /var/folders/2q/qhdc5tpx4knd4pjr1hdskmjc0000gn/T/beaker20150901-9475-12lsc14 agent:/tmp/apply_manifest.pp.Fqrk8n {:ignore => }

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) 15:59:51$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.Fqrk8n
Notice: Compiled catalog for ivvxvftriad6t2h.delivery.puppetlabs.net in environment production in 0.70 seconds
Info: Applying configuration version '1441148392'
Notice: /Stage[main]/Main/Package[apache2]/ensure: created
Notice: Applied catalog in 17.48 seconds

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) executed in 19.38 seconds
Exited: 2

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) 16:00:10$ chkconfig --del httpd
bash: chkconfig: command not found

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) executed in 0.01 seconds
Exited: 127

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) 16:00:10$ ln -s /etc/ /etc/rc.d

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) executed in 0.01 seconds

  * ensure enabling service creates the start & kill symlinks

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) 16:00:10$ mktemp -t apply_manifest.pp.XXXXXX
/tmp/apply_manifest.pp.rhcp42

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) executed in 0.01 seconds
localhost $ scp /var/folders/2q/qhdc5tpx4knd4pjr1hdskmjc0000gn/T/beaker20150901-9475-w84cmz agent:/tmp/apply_manifest.pp.rhcp42 {:ignore => }

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) 16:00:10$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.rhcp42
Notice: Compiled catalog for ivvxvftriad6t2h.delivery.puppetlabs.net in environment production in 0.55 seconds
Info: Applying configuration version '1441148412'
Error: Could not enable apache2:
Error: /Stage[main]/Main/Service[apache2]/enable: change from false to true failed: Could not enable apache2:
Notice: Applied catalog in 0.31 seconds

ivvxvftriad6t2h.delivery.puppetlabs.net (agent) executed in 2.07 seconds
Exited: 4
```

**Comment by er0ck** [comment]
looks like debian8, by default, uses sysV :dizzy_face: , which i didn't expect...  
but after fixing the test to expect this it appears we have a legit fail:
```
root@ycn3o98zi9wix82:~# find  /etc/ -name S02apache2
/etc/rc5.d/S02apache2
/etc/rc4.d/S02apache2
/etc/rc3.d/S02apache2
/etc/rc2.d/S02apache2
root@ycn3o98zi9wix82:~# puppet resource service apache2
service { 'apache2':
  ensure => 'stopped',
}
```
it's not enabled but the start links persist...

**Comment by joshcooper** [comment]
Yay for tests! Maybe we mark the Debian8 as pending and file a separate ticket to fix the provider?

**Comment by er0ck** [comment]
on it... got it to pass sles12.  working on vivid, which should be systemd.

**Comment by er0ck** [comment]
ubuntu15 looks legit also: (outside of the test)
```
root@bnyr5s806rfghrh:~# puppet resource service apache2 enable=true
Error: Could not enable apache2:
Error: /Service[apache2]/enable: change from false to true failed: Could not enable apache2:
service { 'apache2':
  ensure => 'stopped',
  enable => 'false',
}
```

**Comment by er0ck** [comment]
ubuntu15 and debian8 fails ticketed at PUP-5149

**Comment by er0ck** [comment]
https://github.com/puppetlabs/puppet/pull/4208

**Comment by er0ck** [comment]
@joshcooper https://github.com/puppetlabs/puppet/pull/4213

**Pull request merged**